Add function raise an exception when no services found or end of list reached

JamesGibo opened the pull request:
So support for exponential backoff can be added to the Node API, when registry's return 5xx

JamesGibo committed:
Fix bug when using a priority of 99

JamesGibo committed:
Add function to get list of valid services

JamesGibo committed:
Fix tests and linting

JamesGibo committed:
Add tests for function to list discovered services

JamesGibo committed:
Version bump

JamesGibo committed:
Test code tidy

JamesGibo committed:
Add support to remove services from cache with priority above 100 when used,
to match what happens with a priority  < 100

JamesGibo committed:
Merge branch 'andrewbo-auth-txt' into jamesg-backoff

andrewbonney commented on mdnsbridge/mdnsbridgeclient.py:
```diff
@@ -76,28 +76,50 @@ def getHref(self, srv_type, priority=None, api_ver=None, api_proto=None):
             if priority >= 100:
                 if service["priority"] == priority:
-                    return self._createHref(service)
+                    return [service]
```
Was the intention not to gather the full list of services matching this fixed priority?

andrewbonney commented on mdnsbridge/mdnsbridgeclient.py:
```diff
@@ -76,28 +76,50 @@ def getHref(self, srv_type, priority=None, api_ver=None, api_proto=None):
+            href_list = []
+            for service in valid_services:
+                href_list.append(self._createHref(service))
+                self.services[srv_type].remove(service)
```
Is there going to be an issue here if an application (for some reason) called both `getHrefList` and `getHref`? It looks like `getHref` may gradually empty `self.services` so that `getHrefList` isn't guaranteed to return the full set. I wonder if `_getServiceList` should have an option to force a call to `_updateServices`, or perhaps the check for whether `self.services` contains any valid entries should only exist within `getHref` rather than inside `_getServiceList`?

JamesGibo committed:
Version bump

JamesGibo committed:
Merge commit '4141a0dd50e76ebd8f66327a3c8829e79cd880af' into jamesg-backoff

JamesGibo committed:
Remove list of hrefs
Add flag to flush cache of services
For priorities over 100 return a list services that match the priority

JamesGibo committed:
Add test to check random high priority service selected

JamesGibo committed:
Fix typo

JamesGibo committed:
Add separate method for getting service with exceptions

JamesGibo committed:
Refactor getHref to raise exceptions when service found or end of list reached

JamesGibo committed:
Add tests for method for getting service href with exceptions

JamesGibo committed:
Make updateServices a public method

JamesGibo commented:
PR finished, will squash and merge.
Separate PR will be raised for concatenating Registration DNS-SD service type results

andrewbonney commented on mdnsbridge/mdnsbridgeclient.py:
```diff
@@ -22,27 +22,35 @@
+            try:
+                return self.getHrefWithException(srv_type, priority, api_ver, api_proto)
+            except EndOfServiceList:
+                self.logger.writeInfo("End of Aggregator list, reloading")
```
```suggestion
                self.logger.writeInfo("End of DNS-SD service list, reloading")
```

JamesGibo replied:
Done

andrewbonney commented on mdnsbridge/mdnsbridgeclient.py:
```diff
@@ -22,27 +22,35 @@
+                # Re-try after cache has been updated
+                return self.getHrefWithException(srv_type, priority, api_ver, api_proto)
+        except NoService:
+            self.logger.writeWarning("No Aggregator for for {}, priority={}, api_ver={}, api_proto={}".format(
```
```suggestion
            self.logger.writeWarning("No DNS-SD service found for {}, priority={}, api_ver={}, api_proto={}".format(
```

JamesGibo replied:
Done

andrewbonney commented on mdnsbridge/mdnsbridgeclient.py:
```diff
@@ -110,7 +120,7 @@ def _createHref(self, service):
         return '{}://{}:{}'.format(proto, address, port)
 
-    def _updateServices(self, srv_type):
+    def updateServices(self, srv_type):
```
Could this be changed back to an underscore? It doesn't seem like it will be necessary to call it externally.

JamesGibo replied:
I am going to use this method in the node aggregator to flush the cache, so simplify the architecture

andrewbonney commented on mdnsbridge/mdnsbridgeclient.py:
```diff
@@ -53,22 +61,35 @@ def getHref(self, srv_type, priority=None, api_ver=None, api_proto=None):
+        href = self._createHref(service)
+        try:
+            self.services[srv_type].remove(service)
+        except Exception:
```
When does this raise an exception?

JamesGibo replied:
Don't think it can any more so will remove. 
Think I had a point where lists were out of sync

andrewbonney commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+
+## 0.8.2
+- Add method to return list of all service available
```
This may need updating given the final design

andrewbonney replied:
Actually, it may count as a minor rather than patch bump given the added method.

andrewbonney reviewed (commented):
Visually looks like it will do the job. Just a few minor notes to look at. Once they're sorted I'll give it a test locally to ensure the existing behaviour remains the same.

JamesGibo committed:
Bump minor version and tidy log messages
